### PR TITLE
Makes it possible to change installed plugin version

### PR DESCRIPTION
**Pull request by vitaliy-guliy**
### What does this PR do?
Displays dropdown with available versions for each plugin. 

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/13913

Screenshot from 2019-08-27 18-09-18

### CODE REVIEW

**Comment by vitaliy-guliy** [comment]
Now the list of available plugins is displayed in reverse order. The latest version displayed the first.

File: extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts [hunk 1]
```diff
@@ -167,52 +168,261 @@ export class ChePluginManager {
...
+        const rawPlugins = await this.chePluginService.getPlugins(this.activeRegistry, filter);
+
+        // group the plugins
+        const grouppedPlugins = this.groupPlugins(rawPlugins);
+
+        // prepare list of installed plugins without versions and repository URI
+        const installedPluginsInfo = this.getInstalledPluginsInfo();
+
+        // update `installed` field for all the plugin
+        // if the plugin is installed, we ned to set the proper version
```
Comment (mmorhun) [inline extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts]: typo `ned` -> `need`

File: extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts [hunk 2]
```diff
@@ -167,52 +168,261 @@ export class ChePluginManager {
...
+        this.installedPlugins.forEach(plugin => {
+            if (plugin.endsWith('/meta.yaml')) {
+                // it's non default registry
+                // we have to remove '/meta.yaml' from the end
+                plugin = plugin.substring(0, plugin.lastIndexOf('/'));
+            }
+
+            const version = plugin.substring(plugin.lastIndexOf('/') + 1);
+
+            // remove the version by deleting all the text withing the last '/' character
```
Comment (mmorhun) [inline extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts]: Typo `withing` -> `after the last /, including the /`

File: extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts [hunk 3]
```diff
@@ -167,52 +168,261 @@ export class ChePluginManager {
...
+            } else if (!uri.endsWith('/')) {
+                uri += '/';
+            }
+
+            registries.push(uri);
+        });
+
+        const plugins: { publisherName: string, version: string }[] = [];
+        this.installedPlugins.forEach(plugin => {
+            if (plugin.endsWith('/meta.yaml')) {
```
Comment (mmorhun) [inline extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts]: What if a user gives full link to the default registry? Will it duplicate versions?

File: extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts [hunk 4]
```diff
@@ -167,52 +168,261 @@ export class ChePluginManager {
...
+        });
+
+        return plugins;
+    }
+
+    /**
+     * Groups all versions of the same plugin in one structure.
+     */
+    private groupPlugins(rawPlugins: ChePluginMetadata[]): ChePlugin[] {
+        const mapa: { [pluginKey: string]: ChePlugin } = {};
```
Comment (mmorhun) [inline extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts]: I would like another name for this.

File: extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts [hunk 5]
```diff
@@ -167,52 +168,261 @@ export class ChePluginManager {
...
+            }
+        }
+
+        return items;
+    }
+
+    /**
+     * Installs the plugin.
+     */
+    async install(cp: ChePlugin): Promise<boolean> {
```
Comment (mmorhun) [inline extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts]: Up to you, but full name is [better](https://github.com/theia-ide/theia/wiki/Coding-Guidelines#whole-words-names).

File: extensions/eclipse-che-theia-plugin-ext/src/node/che-plugin-service.ts [hunk 6]
```diff
@@ -383,7 +351,7 @@ export class ChePluginServiceImpl implements ChePluginService {
                 if (foundIndex >= 0) {
                     plugins.splice(foundIndex, 1);
                 } else {
-                    workspace.devfile.components.splice(workspace.devfile.components.indexOf(component), 1);
+                    workspace.devfile!.components!.splice(workspace.devfile!.components!.indexOf(component), 1);
```
Comment (mmorhun) [inline extensions/eclipse-che-theia-plugin-ext/src/node/che-plugin-service.ts]: Could you please explain to me why `components` always exist? (It is more about the ` workspace.devfile.components!.forEach((component: cheApi.workspace.devfile.Component) => {` line above (342-nd at the moment of review).

**Comment by vitaliy-guliy** [comment]
@AndrienkoAleksandr @mmorhun  I will prepare another PR  solving your feedbacks. I occasionally merged this one immediately I saw your approve.

File: extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts (hunk 1 above)
Comment (vitaliy-guliy) [reply extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts]: Thanks

File: extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts (hunk 2 above)
Comment (vitaliy-guliy) [reply extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts]: Thanks

File: extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts (hunk 3 above)
Comment (vitaliy-guliy) [reply extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts]: Thanks for pointing me on this case. I did a note that I have to check the registry URI and don't allow the user to add already existing registry.

File: extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts (hunk 4 above)
Comment (vitaliy-guliy) [reply extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts]: Changed on `pluginMap`

File: extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts (hunk 5 above)
Comment (vitaliy-guliy) [reply extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts]: Done

File: extensions/eclipse-che-theia-plugin-ext/src/node/che-plugin-service.ts (hunk 6 above)
Comment (vitaliy-guliy) [reply extensions/eclipse-che-theia-plugin-ext/src/node/che-plugin-service.ts]: I added a verification for the `components` and its initialization if it's no exist.

**Comment by vitaliy-guliy** [comment]
@mmorhun @AndrienkoAleksandr I created next one containing fixes https://github.com/eclipse/che-theia/pull/418

**Pull request merged**